run/dag/dag.go

```diff
@@ -227,33 +232,82 @@ func (d *DAG[V]) HasCycle(id ID) bool {
 
 // Order returns the topological order of the DAG. The node ids are
 // lexicographic sorted whenever possible to give a consistent output.
+// May only be called after Validate().
 func (d *DAG[V]) Order() []ID {
-	order := []ID{}
-	visited := Visited{}
-	for _, id := range d.IDs() {
-		if _, ok := visited[id]; ok {
-			continue
+	if !d.validated {
+		panic(errors.E(errors.ErrInternal, "please report this as a bug"))
+	}
+	return d.order
+}
+
+// calcTopologicalOrder does topological ordering based on Kahns algorithm
+func (d *DAG[V]) calcTopologicalOrder() {
+	// Helper maps to lookup predecessors and successors of a node.
+	// Roots of the grph are nodes with no predecessors.
+	predMap := map[ID]map[ID]struct{}{}
+	succMap := map[ID]map[ID]struct{}{}
+
+	// Populate maps with data from the dag.
+	for id := range d.dag {
+		predMap[id] = map[ID]struct{}{}
+		succMap[id] = map[ID]struct{}{}
+	}
+	for id, preds := range d.dag {
+		for _, predid := range preds {
+			predMap[id][predid] = struct{}{}
+			succMap[predid][id] = struct{}{}
 		}
-		d.walkFrom(id, func(id ID) {
-			if _, ok := visited[id]; !ok {
-				order = append(order, id)
-			}
+	}
+
+	// Will contain the topological order.
+	order := []ID{}
 
-			visited[id] = struct{}{}
-		})
+	// Temporary set that contains nodes without predecessors for the algorithm.
+	// Will be expanded as we remove edges.
+	s := []ID{}
```

Comment:
As I read, it may be better to rename the variable to queue for better readability?
```suggestion
	queue := []ID{}
```

Reply:
Agreed, done in [e623fed](https://github.com/terramate-io/terramate/pull/2165/commits/e623fed5c4b669bf1da740756ede97c6f733374a).